File: commons/src/main/java/l2s/commons/net/netty/SendablePacket.java
package l2s.commons.net.netty;

import io.netty.buffer.ByteBuf;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.nio.charset.StandardCharsets;

public abstract class SendablePacket {
    protected static final Logger LOGGER = LogManager.getLogger(SendablePacket.class);

    protected void writeString(ByteBuf byteBuf, CharSequence charSequence) {
        byteBuf.writeShort(charSequence.length());
        byteBuf.writeCharSequence(charSequence, StandardCharsets.UTF_8);
    }

    public ByteBuf write(ByteBuf byteBuf) {
        try {
            byteBuf.writeByte(getOpCode());
            return writeImpl(byteBuf);
        } catch (RuntimeException e) {
            LOGGER.error("", e);
        }

        return null;
    }

    protected abstract ByteBuf writeImpl(ByteBuf byteBuf);

    protected abstract byte getOpCode();
}

File: commons/src/main/java/l2s/commons/net/netty/SendablePacket.kt
package l2s.commons.net.netty

import com.google.common.flogger.FluentLogger
import io.netty.buffer.ByteBuf
import java.nio.charset.StandardCharsets

/**
 * @author Java-man
 */
abstract class SendablePacket {

    protected fun writeString(byteBuf: ByteBuf, charSequence: CharSequence) {
        byteBuf.writeShort(charSequence.length)
        byteBuf.writeCharSequence(charSequence, StandardCharsets.UTF_8)
    }

    fun write(byteBuf: ByteBuf): ByteBuf {
        try {
            byteBuf.writeByte(opCode.toInt())
            return writeImpl(byteBuf)
        } catch (e: RuntimeException) {
            logger.atSevere().withStackTrace(com.google.common.flogger.StackSize.FULL).withCause(e).log("Can't write buffer.")
            throw RuntimeException("Can't write buffer.", e)
        }
    }

    protected abstract val opCode: Byte

    protected abstract fun writeImpl(byteBuf: ByteBuf): ByteBuf

    companion object {

        @JvmStatic
        protected val logger = FluentLogger.forEnclosingClass()

    }

}